src/brod_gssapi.erl

```diff
@@ -5,134 +5,110 @@
-                        send_sasl_token(Token, Sock, Mod),
-                        SaslRes;
-                    true ->
-                        SaslRes
-                end
-                          end,
-            SaslRes =
-                case do_while(StartCliFun, CondFun) of
-                    SomeRes when SomeRes /= ?SASL_OK andalso SomeRes /= ?SASL_CONTINUE ->
-                        {error, SomeRes};
-                    Other ->
-                        Other
-                end,
-            case SaslRes of
-                ?SASL_OK ->
-                    ok = setopts(Sock, Mod, [{active, once}]);
-                ?SASL_CONTINUE ->
-                    sasl_recv(Mod, Sock, Timeout)
-            end;
-        Other ->
-            {error, Other}
-    end.
-
-sasl_recv(Mod, Sock, Timeout) ->
-    case Mod:recv(Sock, 4, Timeout) of
-        {ok, <<0:32>>} ->
-            ok = setopts(Sock, Mod, [{active, once}]);
-        {ok, <<BrokerTokenSize:32>>} ->
-            case Mod:recv(Sock, BrokerTokenSize, Timeout) of
-                {ok, BrokerToken} ->
-                    CondFun = fun(?SASL_INTERACT) -> continue; (Other) -> Other end,
-                    CliStepFun = fun() ->
-                        {SaslRes, Token} = sasl_auth:sasl_client_step(BrokerToken),
-                        if
-                            SaslRes >= 0 ->
-                                send_sasl_token(Token, Sock, Mod),
-                                SaslRes;
-                            true ->
-                                SaslRes
-                        end
-                                 end,
-                    case do_while(CliStepFun, CondFun) of
-                        ?SASL_OK ->
-                            ok = setopts(Sock, Mod, [{active, once}]);
-                        ?SASL_CONTINUE ->
-                            sasl_recv(Mod, Sock, Timeout);
-                        Other ->
-                            {error, Other}
-                    end
-            end;
-        {error, closed} ->
-            {error, bad_credentials};
-        Unexpected ->
-            {error, Unexpected}
-    end.
+% Observed Sequence of handshake is as follows:
+-spec auth(
+    Host :: string(),
+    Sock :: gen_tcp:socket() | ssl:sslsocket(),
+    HandshakeVsn :: non_neg_integer(),
+    Mod :: gen_tcp | ssl,
+    ClientId :: binary(),
+    Timeout :: pos_integer(),
+    SaslOpts :: term()
+) -> ok | {error, Reason :: term()}.
+auth(
+    Host,
+    Sock,
+    HandshakeVsn,
+    Mod,
+    ClientId,
+    Timeout,
+    Opts
+) ->
+    State = new(Host, Sock, HandshakeVsn, Mod, ClientId, Timeout, Opts),
+    dispatch(State).
 
-%%====================================================================
-%% Internal functions
-%%====================================================================
+dispatch(#{handshake_vsn := 1} = State) ->
+    brod_gssapi_v1:auth(State);
+dispatch(#{handshake_vsn := 0} = State) ->
```

Comment:
When is the the `handshake_vsn = 0`?

Reply:
This would be passed in from kafka_protocol, but it's probably  better as `undefined` since there is no handshake `0`.  Thanks for calling it out :)

Reply:
@zmstone Can you comment on the above? Either way, I'll have to get this merged tomorrow.

Reply:
Actulally, here's what I'll do, I'll go ahead and merge this and open a few issues :)

Reply:
https://github.com/kafka4beam/brod_gssapi/issues/8

Reply:
I don't have any strong feeling either way. Perhaps @zmstone also wants to comment on this as he has much more experience with Kafka.